source/mir/random/tinflex/internal/area.d

```diff
@@ -0,0 +1,650 @@
+in
+{
+    assert(iv.lx < iv.rx, "invalid interval");
+}
+body
+{
+    import mir.random.tinflex.internal.linearfun : emptyFun, secant, tangent;
+    import mir.random.tinflex.internal.types : determineType, FunType;
+
+    enum sec = "secant(iv.lx, iv.rx, iv.ltx, iv.rtx)";
+    enum t_l = "tangent(iv.lx, iv.ltx, iv.lt1x)";
+    enum t_r = "tangent(iv.rx, iv.rtx, iv.rt1x)";
+
+    // could potentially be saved for subsequent calls
+    FunType type = determineType(iv);
+    with(FunType) with(iv)
+    switch(type)
+    {
+        case T1a:
+            squeeze = mixin(t_r);
+            hat = mixin(t_l);
+            break;
+        case T1b:
+            squeeze = mixin(t_l);
+            hat = mixin(t_r);
+            break;
+        case T2a:
+            squeeze = mixin(sec);
+            hat = mixin(t_l);
+            break;
+        case T2b:
+            squeeze = mixin(sec);
+            hat = mixin(t_r);
+            break;
+        case T3a:
+            squeeze = mixin(t_r);
+            hat = mixin(sec);
+            break;
+        case T3b:
+            squeeze = mixin(t_l);
+            hat = mixin(sec);
+            break;
+        case T4a:
+            if (iv.lx == -S.infinity)
+            {
+                squeeze = squeeze.init;
+                hat = mixin(t_r);
+                break;
+            }
+            if (iv.rx == +S.infinity)
+            {
+                squeeze = squeeze.init;
+                hat = mixin(t_l);
+                break;
+            }
+            squeeze = mixin(sec);
+            hat = iv.ltx > iv.rtx ? mixin(t_l) : mixin(t_r);
+            break;
+        case T4b:
+            squeeze = iv.ltx < iv.rtx ? mixin(t_l) : mixin(t_r);
+            hat = mixin(sec);
+            break;
+        default:
+            squeeze = emptyFun!S;
+            hat = emptyFun!S;
+    }
+}
+
+// TODO: add more tests
+unittest
+{
+    import std.meta : AliasSeq;
+    import mir.random.tinflex.internal.types: determineType;
+    foreach (S; AliasSeq!(float, double, real))
+    {
+        const f0 = (S x) => x * x;
+        const f1 = (S x) => 2 * x;
+        const f2 = (S x) => 2.0;
+        auto c = 42; // not required for this test
+        auto dhs = (S l, S r) {
+            auto iv = Interval!S(l, r, c, f0(l), f1(l), f2(l),
+                                          f0(r), f1(r), f2(r));
+            determineSqueezeAndHat(iv);
+            return iv;
+        };
+
+        // test left side
+        auto hs1 = dhs(-1, 1);
+        assert(hs1.hat.slope == 0);
+        assert(hs1.hat.intercept == 1);
+        assert(hs1.squeeze.slope == 2);
+        assert(hs1.squeeze.intercept == -1);
+
+        // test right side
+        auto hs2 = dhs(1, 3);
+        assert(hs2.hat.slope == 4);
+        assert(hs2.hat.intercept == -3);
+        assert(hs2.squeeze.slope == 2);
+        assert(hs2.squeeze.intercept == -1);
+    }
+}
+
+alias hatArea(S) = area!(true, S);
+alias squeezeArea(S) = area!(false, S);
```

Comment:
> OK, in that case they should be marked private (or package if they're needed by other tinflex-related modules).

@9il's policy is to make everything public for hackers, but state with `internal` that the API isn't supported

@9il can I move them back down so that the aliases are closer to the `area` function?